### ISSUE TITLE
Add `user-agent` to default list of `Access-Control-Allow-Headers`

### ISSUE THREAD

**Issue by sohkai** [ipfs/go-ipfs#5138]
#### Version information:

**Mostly to do with public gateways**, but issue persists on recent versions:

```
go-ipfs version: 0.4.13-
Repo version: 6
System version: amd64/darwin
Golang version: go1.9.2
```

#### Type:
Enhancement

#### Description:

When sending requests across CORS, it looks like newer browsers are asking for `user-agent` in `Access-Control-Allow-Headers`, and failing fetches without it, e.g.:

![image](https://user-images.githubusercontent.com/4166642/41591398-df35590c-73b9-11e8-8725-9b0d1669602a.png)

It would be really nice if `user-agent` was included by default in the CORS settings. Also, not sure if setting it twice is the same as a list, but it seems like it [should be a list](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers)?

[See a similar API issue (Github's)](https://github.com/octokit/rest.js/issues/817)

**Comment by bpierre** [comment]
In the meantime, it should be possible to update the headers using the config file: https://github.com/ipfs/go-ipfs/blob/7e8f6c96047e1a94c1c892aa50e1c75e01b7fda2/docs/config.md#gateway

**Comment by ivan386** [comment]
@bpierre Yes it possible but this setting must be set on all gateways by default. 

Firefox block request by XMLHttpRequest to all public gateways without it.

**Comment by lidel** [comment]
Big yes for allowing for `user-agent` customization during `fetch`, we want to be compatible with vendors following the spec (https://github.com/whatwg/fetch/issues/37) :+1:

As for the cleanup of defaults:

**tl;dr** I think we should go with (2) AND while at it make sure we will be able to do some cleanup described below.

----

I've looked at our headers recently and started tracking related issues at https://github.com/ipfs/in-web-browsers/issues/132.

As you can see in "go-ipfs v0.4.18 defaults" section of  https://github.com/ipfs/in-web-browsers/issues/132#issuecomment-442039654 we already have "hardcoded" defaults for headers such as `Access-Control-Expose-Headers`.  

The need to hardcode support for `user-agent` out of the box  in `-Allow-Headers`  is similar to having `X-Chunked-Output` in `-Expose-Headers` (go-ipfs returns it even if not present in config because "it is part of the stack" and without it we are unable to  [detect streaming responses in js-ipfs-api](https://github.com/ipfs/js-ipfs-api/blob/52298ae1bf563a1183b3fce86353cc009f19569e/src/utils/send-request.js#L35-L36)).


@Stebalien  Note that [`-Allow-Headers` and `-Expose-Headers`](https://github.com/ipfs/in-web-browsers/issues/132#issuecomment-442046210) control different things (request vs response) and defaults for each should be hardcoded separately, so we can clean them up later (eg. `X-Chunked-Output` makes sense only in  `-Expose-Headers`).
